File: src/commands/push.ts
import {Args, Flags} from '@oclif/core'
import type {ResolvedSubrepo} from '../config.js'
import {MonospliceCommand} from '../lib/base.js'
import {
  DRY_RUN_NOTE,
  confirmFirstPublish,
  exportSubrepo,
  firstPublish,
  loadView,
  planPushDryRun,
  upstreamHasNoBranch,
  type Reporter,
} from '../lib/ops.js'

interface PushFlags {
  yes: boolean
  'export-history': boolean
  'dry-run': boolean
}

export default class Push extends MonospliceCommand {
  static description = 'Export new monorepo commits to the standalone subrepo remotes'

  static args = {
    subrepo: Args.string({description: 'Only push this subrepo (defaults to all)', required: false}),
  }

  static flags = {
    yes: Flags.boolean({
      char: 'y',
      description: 'Answer the first-publish confirmation with yes (required in scripts and CI)',
      default: false,
    }),
    'export-history': Flags.boolean({
      description:
        'First publish only: replay every monorepo commit that touched the subrepo instead of one baseline commit (not to be confused with `attach --import-history`, which replays the standalone repo\'s commits inwards)',
      default: false,
    }),
    'dry-run': Flags.boolean({
      description:
        'List the commits a push would export and write nothing — no remote ref, no commit, no working-tree change. Scan/transform hooks do NOT run on a dry run, so the list is what would be attempted; the hooks still gate the real push and a rejected commit will stop it.',
      default: false,
    }),
  }

  static examples = [
    '<%= config.bin %> <%= command.id %>',
    '<%= config.bin %> <%= command.id %> core',
    '<%= config.bin %> <%= command.id %> --dry-run',
    '<%= config.bin %> <%= command.id %> core --yes',
    '<%= config.bin %> <%= command.id %> core --yes --export-history',
  ]

  async run(): Promise<void> {
    const {args, flags} = await this.parse(Push)
    const project = await this.requireProject()

    // One subrepo refusing (typically: never published, no --yes) must not silence the
    // others, so failures are collected and reported together at the end.
    const reporter = this.collectingReporter()
    const selected = this.selectSubrepos(project, args.subrepo)

    if (flags['dry-run']) {
      await this.eachSubrepo(selected, (subrepo) => this.previewOne(project.root, subrepo, reporter, flags))
      return
    }

    await this.eachSubrepo(selected, (subrepo) => this.pushOne(project.root, subrepo, reporter, flags))
  }

  /** Report the plan and stop. Every call below this line is a read. */
  private async previewOne(
    root: string,
    subrepo: ResolvedSubrepo,
    r: Reporter,
    flags: PushFlags,
  ): Promise<void> {
    const plan = await planPushDryRun(root, subrepo, r, {exportHistory: flags['export-history']})

    if (plan.kind === 'first-publish') {
      const how = plan.exportHistory
        ? `replaying ${plan.commits.length} commit(s)`
        : 'one baseline commit'
      this.log(
        `${subrepo.name}: would publish ${subrepo.path}/ to ${subrepo.remote} (${subrepo.branch}) for the first time — ${how} (${DRY_RUN_NOTE})`,
      )
    } else if (plan.commits.length === 0) {
      this.log(`${subrepo.name}: up to date (${DRY_RUN_NOTE})`)
      return
    } else {
      this.log(`${subrepo.name}: ${plan.commits.length} to push (${DRY_RUN_NOTE})`)
    }

    for (const c of plan.commits) this.log(`  ${c.sha.slice(0, 10)} ${c.subject}`)
  }

  private async pushOne(
    root: string,
    subrepo: ResolvedSubrepo,
    r: Reporter,
    flags: PushFlags,
  ): Promise<void> {
    const view = await loadView(root, subrepo, r)

    if (view.pubHead === null && subrepo.upstream !== undefined) r.fail(upstreamHasNoBranch(subrepo))

    if (view.pubHead === null) {
      const result = await firstPublish(root, subrepo, r, {
        exportHistory: flags['export-history'],
        confirm: () => confirmFirstPublish(subrepo, r, {yes: flags.yes}),
      })
      const how = result.exportHistory ? `replayed ${result.commits} commit(s)` : 'one baseline commit'
      this.log(`✓ ${subrepo.name}: published ${subrepo.path}/ to ${subrepo.remote} (${subrepo.branch}) — ${how}`)
      return
    }

    if (flags['export-history']) {
      r.fail(
        `${subrepo.name}: --export-history only applies to the first publish, and ${subrepo.remote} already has a ${subrepo.branch} branch (${view.pubHead.slice(0, 10)}).
Nothing was pushed. Run \`monosplice push ${subrepo.name}\` to export new commits.`,
      )
    }

    const {pushed, awaiting} = await exportSubrepo(root, subrepo, r, view)
    const fork = subrepo.upstream === undefined ? '' : ` to ${subrepo.remote} (${subrepo.pushBranch})`

    if (pushed > 0) this.log(`✓ ${subrepo.name}: exported ${pushed} commit(s)${fork}`)
    else if (awaiting > 0) {
      this.log(
        `✓ ${subrepo.name}: up to date — ${subrepo.remote} (${subrepo.pushBranch}) already carries ${awaiting} commit(s), awaiting an upstream merge`,
      )
    } else this.log(`✓ ${subrepo.name}: up to date`)
  }
}

File: src/commands/push.rs
//! Port of the corresponding TypeScript module — see docs/rust-port.md.
//!
//! Export new monorepo commits to the standalone remotes. One subrepo refusing (typically:
//! never published, no `--yes`) must not silence the others, so failures are collected by
//! [`each_subrepo`] and reported together at the end.

use std::path::Path;

use crate::config::ResolvedSubrepo;
use crate::core::sync_view::SyncViewOptions;
use crate::ops::{
    confirm_first_publish, export_subrepo, first_publish, load_view, plan_push_dry_run, short,
    upstream_has_no_branch, ConfirmFirstPublishOptions, DryRunPlan, DRY_RUN_NOTE,
};
use crate::report::{each_subrepo, require_project, select_subrepos, Failure, SubrepoFailure};

#[derive(clap::Args, Debug)]
pub struct PushArgs {
    #[arg(
        value_name = "subrepo",
        help = "Only push this subrepo (defaults to all)"
    )]
    pub subrepo: Option<String>,

    #[arg(
        short = 'y',
        long,
        help = "Answer the first-publish confirmation with yes (required in scripts and CI)"
    )]
    pub yes: bool,

    #[arg(
        long = "export-history",
        help = "First publish only: replay every monorepo commit that touched the subrepo instead of one baseline commit (not to be confused with `attach --import-history`, which replays the standalone repo's commits inwards)"
    )]
    pub export_history: bool,

    #[arg(
        long = "dry-run",
        help = "List the commits a push would export and write nothing — no remote ref, no commit, no working-tree change. Scan/transform hooks do NOT run on a dry run, so the list is what would be attempted; the hooks still gate the real push and a rejected commit will stop it."
    )]
    pub dry_run: bool,
}

pub fn run(args: &PushArgs) -> Result<(), Failure> {
    let project = require_project()?;
    let selected = select_subrepos(&project, args.subrepo.as_deref())?;
    let root = project.root.as_path();

    if args.dry_run {
        return each_subrepo(&selected, |subrepo| preview_one(root, subrepo, args));
    }

    each_subrepo(&selected, |subrepo| push_one(root, subrepo, args))
}

/// Report the plan and stop. Every call below this line is a read.
fn preview_one(
    root: &Path,
    subrepo: &ResolvedSubrepo,
    args: &PushArgs,
) -> Result<(), SubrepoFailure> {
    let plan = plan_push_dry_run(root, subrepo, args.export_history)?;

    if let DryRunPlan::FirstPublish {
        export_history,
        commits,
    } = &plan
    {
        let how = if *export_history {
            format!("replaying {} commit(s)", commits.len())
        } else {
            "one baseline commit".to_string()
        };
        println!(
            "{}: would publish {}/ to {} ({}) for the first time — {how} ({DRY_RUN_NOTE})",
            subrepo.name, subrepo.path, subrepo.remote, subrepo.branch
        );
    } else if plan.commits().is_empty() {
        println!("{}: up to date ({DRY_RUN_NOTE})", subrepo.name);
        return Ok(());
    } else {
        println!(
            "{}: {} to push ({DRY_RUN_NOTE})",
            subrepo.name,
            plan.commits().len()
        );
    }

    for c in plan.commits() {
        println!("  {} {}", short(&c.sha), c.subject);
    }
    Ok(())
}

fn push_one(root: &Path, subrepo: &ResolvedSubrepo, args: &PushArgs) -> Result<(), SubrepoFailure> {
    let view = load_view(root, subrepo, SyncViewOptions::default())?;

    if view.pub_head.is_none() && subrepo.upstream.is_some() {
        return Err(SubrepoFailure::new(upstream_has_no_branch(subrepo)));
    }

    if view.pub_head.is_none() {
        let result = first_publish(root, subrepo, args.export_history, || {
            confirm_first_publish(
                subrepo,
                &ConfirmFirstPublishOptions {
                    yes: args.yes,
                    ..Default::default()
                },
            )
        })?;
        let how = if result.export_history {
            format!("replayed {} commit(s)", result.commits)
        } else {
            "one baseline commit".to_string()
        };
        println!(
            "✓ {}: published {}/ to {} ({}) — {how}",
            subrepo.name, subrepo.path, subrepo.remote, subrepo.branch
        );
        return Ok(());
    }

    if args.export_history {
        let pub_head = view.pub_head.as_deref().unwrap_or_default();
        return Err(SubrepoFailure::new(format!(
            "{}: --export-history only applies to the first publish, and {} already has a {} branch ({}).
Nothing was pushed. Run `monosplice push {}` to export new commits.",
            subrepo.name,
            subrepo.remote,
            subrepo.branch,
            short(pub_head),
            subrepo.name
        )));
    }

    let summary = export_subrepo(root, subrepo, Some(view))?;
    let fork = if subrepo.upstream.is_none() {
        String::new()
    } else {
        format!(" to {} ({})", subrepo.remote, subrepo.push_branch)
    };

    if summary.pushed > 0 {
        println!(
            "✓ {}: exported {} commit(s){fork}",
            subrepo.name, summary.pushed
        );
    } else if summary.awaiting > 0 {
        println!(
            "✓ {}: up to date — {} ({}) already carries {} commit(s), awaiting an upstream merge",
            subrepo.name, subrepo.remote, subrepo.push_branch, summary.awaiting
        );
    } else {
        println!("✓ {}: up to date", subrepo.name);
    }
    Ok(())
}
